### ISSUE TITLE
Incorrect fermionic anticommutation relationships

### ISSUE THREAD

**Issue by acroscarrillo** [qojulia/QuantumOpticsBase.jl#102]
I'm fairly new to the package but it seems to me like the [fermionic anticommutation relations](https://en.wikipedia.org/wiki/Creation_and_annihilation_operators#Creation_and_annihilation_operators_in_quantum_field_theories)
```math
\{a_i, a^\dagger_j\}\doteq a_i a^\dagger_j + a^\dagger_j a_i = \delta_{ij} \quad \text{and} \quad \{a_i^\dagger, a^\dagger_j\} = \{a_i, a_j\} = 0,
```
have been incorrectly implemented:
```
julia> b = NLevelBasis(2)
NLevel(N=2)

julia> b_mb = ManyBodyBasis(b, fermionstates(b, [0,1,2]))
ManyBody(onebodybasis=NLevel(N=2), states:4)

julia> (create(b_mb,2)*create(b_mb,1)+create(b_mb,1)*create(b_mb,2))*basisstate(b_mb, [0,0])
Ket(dim=4)
  basis: ManyBody(onebodybasis=NLevel(N=2), states:4)
 0.0 + 0.0im
 0.0 + 0.0im
 0.0 + 0.0im
 2.0 + 0.0im
```
Since, if the code does what I think it should be doing, the last operation **should be** 
```math
(a^\dagger_2 a^\dagger_1 + a^\dagger_1 a^\dagger_2 )|0,0\rangle = ( a^\dagger_2 a^\dagger_1 - a^\dagger_2 a^\dagger_1 )|0,0\rangle = 0
```
but instead it is (wrongly) doing 
```math
(a^\dagger_2 a^\dagger_1 + a^\dagger_1 a^\dagger_2 )|0,0\rangle = 2* a^\dagger_2 a^\dagger_1 |0,0\rangle = 2*|1,1\rangle
```
as if they were bosons. Interestingly:
```
julia> (destroy(b_mb,1)*create(b_mb,1)+create(b_mb,1)*destroy(b_mb,1))*basisstate(b_mb, [0,0])
Ket(dim=4)
  basis: ManyBody(onebodybasis=NLevel(N=2), states:4)
 1.0 + 0.0im
 0.0 + 0.0im
 0.0 + 0.0im
 0.0 + 0.0im
```
which is correct but looks like a fluke judging the above issue.

**Comment by amilsted** [comment]
I think both results are correct for bosons. I think the only thing `fermionstates` does is implement "pauli exclusion", so disallowing >1 fermion in each mode. This is understandably misleading. Perhaps @david-pl can correct me?

**Comment by david-pl** [comment]
I admit to being a bit rusty myself when it comes to `ManyBodyBasis`. But yes, @amilsted is correct, `fermionstates` simply limits the number of particles per site (per level of `NLevelBasis(2)` in the above example) to 1.

With `create(b_mb, 1) * create(b_mb, 2)` you're creating one fermion that is in state `1` and another one that is in state `2`. Since the particle number is not conserved, this is fine and the order of the operations doesn't matter. So the anti-commutation relation isn't applicable here.

**Comment by acroscarrillo** [comment]
Hey thank you both for chipping in!

> With `create(b_mb, 1) * create(b_mb, 2)` you're creating one fermion that is in state 1 and another one that is in state 2. 

The above statement is true, however

> Since the particle number is not conserved, this is fine and the order of the operations doesn't matter. So the anti-commutation relation isn't applicable here.

this is not true: fermionic operator anticommute at different sites regardless of any symmetry of the state. That is $c_1 c_2 = - c_2 c_1$ is true at the level of the underlying algebra.

**Comment by aryavorskiy** [comment]
Okay, I may have some ideas on how to implement this, but this might be a breaking change, so RFC:

Let us define the filling number states as follows:
```math
| N_1^{i_1} N_2^{i_2} \ldots \rangle = 
(\hat{a}^\dagger_{i_1})^{N_1} (\hat{a}^\dagger_{i_2})^{N_2} \ldots |0\rangle, \\ \\ \\
i_1 < i_2 < \ldots
```
The ordering of the creation operators here is unimportant for bosons, but extremely important for fermions. Also for fermions we must have `N_j = 1`, but this is already taken care of by the "Pauli exclusion" that is already implemented.

Thus, if we want to find the result of the creation operator, we can do the following:
```math
\hat{a}^\dagger_j | 1^{i_1} 1^{i_2} \ldots \rangle = 
\hat{a}^\dagger_j \hat{a}^\dagger_{i_1} \hat{a}^\dagger_{i_2} \ldots |0\rangle
= (-1)^k \hat{a}^\dagger_{i_1} \ldots \hat{a}^\dagger_{i_k} \hat{a}^\dagger_j \hat{a}^\dagger_{i_{k+1}} \ldots |0\rangle, \\ \\ \\
i_k < j < i_{k+1}
```

The result of the annihilation operator can be carried out in a similar way.

The simplest way of somehow setting the particle statistics is adding a `isbosestatistics::Bool` field to the `ManyBodyBasis` struct. However, we can implement [Abelian anyons](https://en.m.wikipedia.org/wiki/Anyon#Abelian_anyons) by remembering the `exp(im * ϴ)` phase instead:

```math
\hat{a}_i \hat{a}_j - e^{i \theta} \hat{a}_j \hat{a}_i = 0, \\ \\ \\
\hat{a}_i \hat{a}^\dagger_j - e^{i \theta} \hat{a}^\dagger_j \hat{a}_i = \delta_{i j}
```

But do we really need that complication?

Will file a PR ASAP if no objections.

**Comment by Krastanov** [comment]
Thank you @acroscarrillo for reporting this and thank you @aryavorskiy for the fix. It is now merged (in #160 ) and a release will be made shortly.
